### PR TITLE
fix: cancel migrate bucket v0.2.4

### PR DESCRIPTION
**Pull request by jingjunLi**
### Description

1) Bug Fix: Handling Bucket Migration
When dealing with bucket migration scenarios involving the sequence of actions: "migrate bucket," "cancel migrate," and then "migrate bucket" again, a critical bug has been identified and resolved. The sequence could lead to unexpected behavior. This fix ensures that the process correctly handles such sequences, maintaining the integrity of the migration process.
2) Bug Fix: Retry Mechanism for Failed Task
In cases where pushing a task to the migrateGVGQueue fails, there was a problem where the failed task wasn't retried. This pull request addresses this issue by implementing a proper retry mechanism. Now, if a task fails to be pushed to the queue, the system will make subsequent attempts to ensure the task is properly added.

### Rationale

* Following the cancellation and successful completion of a bucket migration, it's now necessary to remove migrate gvgs from the database. This enhancement maintains a clean and accurate record of migrations.
* The cancelMigrateBucket function has been encapsulated to improve its overall structure and maintainability.
* An issue related to inserting duplicate entries in InsertMigrateGVGUnit has been resolved, ensuring the correct insertion of data.
* The DeleteMigrateGVGUnit function has been introduced to properly handle the deletion of migrate Global View Generators (gvgs).
* To prevent potential concurrency issues, a mutex has been added to the migrateGVGQueue.

### Example

NA

### Changes

Notable changes: 
* NA
* ...

### CODE REVIEW

File: modular/manager/bucket_migrate_scheduler.go [hunk 1]
```diff
@@ -421,28 +421,35 @@ func (s *BucketMigrateScheduler) doneMigrateBucket(bucketID uint64) error {
...
+	expected, err := s.checkBucketFromChain(bucketID, storagetypes.BUCKET_STATUS_CREATED)
+	if err != nil {
+		return err
+	}
+	if !expected {
+		return nil
+	}
+	executePlan, err := s.getExecutePlanByBucketID(bucketID)
+	if err != nil {
+		log.Errorw("bucket migrate schedule received EventCompleteMigrationBucket", "bucket_id", bucketID)
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:438]: Refine this log.

File: modular/manager/manage_task.go [hunk 2]
```diff
@@ -651,20 +651,20 @@ func (m *ManageModular) HandleMigrateGVGTask(ctx context.Context, task task.Migr
 		log.CtxErrorw(ctx, "failed to handle migrate gvg due to pointer dangling")
 		return ErrDanglingTask
 	}
-	m.migrateGVGQueue.PopByKey(task.Key())
 	var err, pushErr error
-	task.SetUpdateTime(time.Now().Unix())
+
+	pushErr = m.migrateGVGQueuePopByLimitAndPushAgain(task)
+	if pushErr != nil {
```
Comment (will-2012) [inline modular/manager/manage_task.go:657]: Print log.

File: modular/manager/manage_task.go [hunk 3]
```diff
@@ -651,20 +651,20 @@ func (m *ManageModular) HandleMigrateGVGTask(ctx context.Context, task task.Migr
...
 		err = m.spExitScheduler.UpdateMigrateProgress(task)
 	}
-	if !task.GetFinished() {
-		if pushErr = m.migrateGVGQueue.Push(task); pushErr != nil {
-			log.CtxErrorw(ctx, "failed to push gvg task queue", "task", task, "error", pushErr)
-		}
-	}
-	log.CtxInfow(ctx, "success to handle migrate gvg task", "task", task, "error", err, "error_push", pushErr)
+
+	log.CtxInfow(ctx, "succeed to handle migrate gvg task", "task", task, "error", err, "error_push", pushErr)
```
Comment (will-2012) [inline modular/manager/manage_task.go:667]: Remove `"err_push", pushErr`.

File: modular/manager/manager.go [hunk 4]
```diff
@@ -746,7 +747,41 @@ func (m *ManageModular) repushTask(reserved task.Task) {
...
+
+	var pushErr error
+
+	m.migrateGVGQueue.PopByKey(task.Key())
+	task.SetUpdateTime(time.Now().Unix())
+	if !task.GetFinished() {
+		if pushErr = m.migrateGVGQueue.Push(task); pushErr != nil {
+			log.Errorw("failed to push gvg task queue", "task", task, "error", pushErr)
+		}
+	}
```
Comment (will-2012) [inline modular/manager/manager.go:784]: Print a trace that includes context(task info ,queue info) to debug.

File: modular/manager/bucket_migrate_scheduler.go [hunk 5]
```diff
@@ -758,6 +765,7 @@ func (s *BucketMigrateScheduler) UpdateMigrateProgress(task task.MigrateGVGTask)
 
 	migrateExecuteUnit, ok := executePlan.gvgUnitMap[gvgID]
 	if !ok {
+		// maybe bucket migrate canceled
```
Comment (will-2012) [inline modular/manager/bucket_migrate_scheduler.go:768]: Print a debug log.

**Review by will-2012 — APPROVED** [review]

File: store/sqldb/migrate.go [hunk 6]
```diff
@@ -292,21 +294,44 @@ func (s *SpDBImpl) InsertMigrateGVGUnit(meta *spdb.MigrateGVGUnitMeta) error {
...
-		result = s.db.Model(&MigrateGVGTable{}).Where("migrate_key = ?", meta.MigrateGVGKey).
-			Updates(insertMigrateGVG)
+	result = s.db.First(queryReturn, "migrate_key = ?", meta.MigrateGVGKey)
+	if result.Error != nil && !errors.Is(result.Error, gorm.ErrRecordNotFound) {
+		return result.Error
 	}
-	if result.Error != nil || result.RowsAffected != 1 {
-		err = fmt.Errorf("failed to insert migrate gvg table: %s", result.Error)
-		return err
+	if result.Error != nil {
```
Comment (sysvm) [inline store/sqldb/migrate.go:301]: Do you add new unit tests to improve UT coverage?

**Review by sysvm — APPROVED** [review]
LGTM

**Pull request merged**